### PR TITLE
Return exit code when exiting with --abort-on-container-exit

### PR DESCRIPTION
**Pull request by kbh2o**
This is taking the code from #3725, which appears to have lost traction and implements the changes requested.

### CODE REVIEW

**Comment by kbh2o** [comment]
bump

**Comment by shin-** [comment]
Thanks!

Looking at Aanand's comment on the original PR (https://github.com/docker/compose/pull/3725#issuecomment-233958678), I think there's one important piece that's missing, with the code as it is currently there is no guarantee that the container that caused the cascade stop is the one we're looking at for the potential exit code change. 

To accomplish that, changes in the `log_printer` might be needed to identify the container that caused the interruption and retrieve that specific code.

**Comment by kbh2o** [comment]
Thanks for the feedback!  I'll look into the `log_printer` and push up changes soon.

**Comment by kbh2o** [comment]
I pushed up some changes that work the suggested way.  Before I modify/write additional tests, I would like to know if this method makes sense.

File: compose/cli/log_printer.py [hunk 1]
```diff
@@ -87,6 +87,12 @@ def run(self):
         for line in consume_queue(queue, self.cascade_stop):
             remove_stopped_threads(thread_map)
 
+            if self.cascade_stop:
+                matching_container = [cont.name for cont in self.containers if cont.name == line]
+                if line in matching_container:
+                    self.cascade_starter = line
```
Comment (shin-) [inline compose/cli/log_printer.py]: maybe `return line` here instead

File: compose/cli/main.py [hunk 2]
```diff
@@ -914,7 +914,13 @@ def up(self, options):
```
Comment (shin-) [inline compose/cli/main.py]: `cascade_starter = log_printer.run()` (can be `None`)

File: compose/cli/log_printer.py (hunk 1 above)
Comment (shin-) [inline compose/cli/log_printer.py]: And add a short comment to explain why we're doing that.

File: compose/cli/log_printer.py [hunk 3]
```diff
@@ -230,7 +235,7 @@ def consume_queue(queue, cascade_stop):
 
         if item.is_stop:
             if cascade_stop:
-                raise StopIteration
+                yield item.item
```
Comment (shin-) [inline compose/cli/log_printer.py]: we already `yield item.item` unconditionally below, so it seems like this whole block can be changed to
```
if item.is_stop and not cascade_stop:
  continue
```

**Comment by kbh2o** [comment]
Thanks for the feedback.  I pushed the changes up.  I have tried to fix signing the "merge from master" commit, but I am not having any luck.

**Comment by shin-** [comment]
Yeah, ideally we don't want merge commits in PRs, they should be rebased against master instead. Not a huge deal. :)

File: tests/fixtures/abort-on-container-exit-0/digest.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+simple:
```
Comment (shin-) [inline tests/fixtures/abort-on-container-exit-0/digest.yml]: File is unused

File: tests/fixtures/abort-on-container-exit-0/ignore-pull-failures.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+simple:
```
Comment (shin-) [inline tests/fixtures/abort-on-container-exit-0/ignore-pull-failures.yml]: File is unused

File: tests/fixtures/abort-on-container-exit-1/digest.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+simple:
```
Comment (shin-) [inline tests/fixtures/abort-on-container-exit-1/digest.yml]: File is unused

File: tests/fixtures/abort-on-container-exit-1/ignore-pull-failures.yml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+simple:
```
Comment (shin-) [inline tests/fixtures/abort-on-container-exit-1/ignore-pull-failures.yml]: File is unused

File: compose/cli/main.py [hunk 8]
```diff
@@ -886,11 +886,17 @@ def up(self, options):
...
-            log_printer.run()
+            cascade_starter = log_printer.run()
 
             if cascade_stop:
                 print("Aborting on container exit...")
+                exit_code = 0
+                for e in self.project.containers(service_names=options['SERVICE'], stopped=True):
+                    if (not e.is_running and cascade_starter == e.name):
+                        if not e.exit_code == 0:
+                            exit_code = e.exit_code
```
Comment (shin-) [inline compose/cli/main.py]: I think we can add a `break` here (inside the if block).

File: compose/cli/main.py [hunk 9]
```diff
@@ -886,11 +886,18 @@ def up(self, options):
...
+                        break
```
Comment (shin-) [inline compose/cli/main.py]: Should indent to be inside the `if not e.exit_code == 0` block

Comment (kbh2o) [inline compose/cli/main.py]: oops. I'll update it.

**Comment by kbh2o** [comment]
not sure what I did... but that doesn't look right... (-.-)

**Comment by shin-** [comment]
Oops! Maybe create a new branch from master `git checkout -b new_branch_name master` and cherry-pick your commits into it. Make sure you've pulled the latest master.

**Comment by kbh2o** [comment]
Created a new PR #4555 I'll close this one out.

**Pull request closed without merging**